### ISSUE TITLE
fix(serve): a deepseek41 GGUF never reaches the native V4.1 forward - the dispatch predicate keys on deepseek_v41, so the serve runs the GLM-DSA MLA forward and dies on the attn_kv width seam

### ISSUE THREAD

**Issue by anthony-chaudhary** [anthony-chaudhary/fak#13246]
# fix(serve): a `deepseek41` GGUF never reaches the native V4.1 forward — the dispatch predicate keys on `deepseek_v41`, so the serve runs the GLM-DSA MLA forward and dies on the `attn_kv` width seam

```routing
lane: model
paths: ["internal/model/v41_config.go", "internal/model/config.go", "internal/ggufload/deepseek41.go", "internal/model/kv.go"]
class: dev
priority: P1
expected_steps: 4
```

centrality: Enabling (the native V4.1 Flash 30/5 physical witness on one Halo appliance)
p1: advanced - the real root cause behind the fak#13245 "shape seam" is named: it is not a projection-width reconciliation, it is a wrong-architecture dispatch. The artifact is correct; the forward selection is wrong.
p2: preserved - every fail-closed guard (`requireResidentShape`, the native v41 admission ladder, the safetensors `refuseDeepSeekV41Native` gate) stays; this issue re-routes to the correct forward rather than relaxing a guard.
p3: preserved - no roofline gate lowered; no checkpoint substituted; the vcruz Q2_K artifact stays the pinned P1 artifact.
p4: preserved - the deepseek2/GLM MLA paths are untouched by the `deepseek41`-identity change.

## Current state

A physical strix3 serve of the exact staged vcruz Q2_K artifact (`58d8ac86`) at public
`9a4cfc973` reaches a fail-closed shape refusal (filed as fak#13245):

```
panic: model: resident k-quant tensor shape mismatch:
  model.layers.0.self_attn.kv_a_proj_with_mqa.weight stored=[512,5120] requested=[576,5120]
```

fak#13245 framed this as "the glm-family MLA forward expects `kvLora+qkRope=576`, but the
artifact's `attn_kv` stores only `kvLora=512`", and proposed reconciling the projection
widths. This issue records the verified evidence that this framing is wrong.

## Verified root cause

The official `deepseek-ai/DeepSeek-V4.1-Flash` reference (`inference/model.py`, pinned
revision `dba1be0a40aa45a94ad051997016db3960a90277`, staged on strix3) does **not** use
latent MLA attention:

```python
class Attention(nn.Module):
    self.head_dim = args.head_dim          # 512
    self.rope_head_dim = args.rope_head_dim  # 64  (qk_rope_head_dim)
    self.nope_head_dim = self.head_dim - self.rope_head_dim
    ...
    self.wkv = Linear(self.dim, self.head_dim)          # [5120 -> 512], NOT low-rank
    self.kv_norm = RMSNorm(self.head_dim, self.eps)     # norm width 512
    ...
    kv = self.kv_norm(self.wkv(x))                      # single KV head, width head_dim
    apply_rotary_emb(kv[..., -self.rope_head_dim :], freqs_cis)  # in-place, LAST 64 only
```

The official `config.json` (`text_config`) carries `head_dim: 512`, `qk_rope_head_dim: 64`,
`q_lora_rank: 1280`, `num_key_value_heads: 1` — and **no `kv_lora_rank` key at all**. There
is no `kv_b_proj`, no separate `k_pe` tensor, and no rope fused into a 576-wide projection.

Therefore the artifact's `blk.<L>.attn_kv.weight [5120, 512]` is **exactly correct**: it is
`wkv`, projecting to one `head_dim=512` KV head. The native V4.1 forward already encodes
this contract (`internal/model/v41_forward.go:869` reads `attn.wkv.weight`; admission at
`:674` requires `[kvLatentRank=512, H]`; rope at `:938`). The 576-wide request is a
GLM-DSA-MLA artifact of running the **wrong forward**.

The dispatch is wrong because the loader's canonical identity and the model-side predicate
disagree:

- `internal/ggufload/gguf_config.go:325-330` maps every V4.1 spelling (including
  `deepseek41`) to the single internal `ModelType == "deepseek41"`.
- `internal/model/v41_config.go:91-93` — `IsDeepSeekV41()` matches only
  `c.DeepSeekV41 != nil || "deepseek_v41" || "deepseek_v41_text"`. A GGUF-loaded config has
  `DeepSeekV41 == nil` and `ModelType == "deepseek41"`, so this is **false**.
- `internal/model/config.go:1109-1111` — `usesMLAMoELayout()` **includes** `"deepseek41"`,
  so it is **true**.
- `internal/model/kv.go:1399` (`Session.Step`) and `:1150` (`Session.Prefill`) test
  `IsDeepSeekV41()` first; it is false, so control falls to `usesMLAMoELayout()` at
  `kv.go:1409`/`:1170`, which calls `tokenHiddenGLMDsa` -> `glmDsaAppendAttentionKV`
  (`internal/model/glm_dsa_session.go:443`) -> the 576-wide request -> the panic.

The current model-side comment at `config.go:1095-1108` claims the vcruz converter "emits
the MLA low-rank attention layout". The reference above shows it does not; the comment,
`deepseek41_mla_route_test.go`, and the fak#13245 framing all encode that mistaken premise.

## Core through-line

Reconcile the loader's canonical naming and the dispatch predicate so a `deepseek41` GGUF
reaches the **native** non-MLA V4.1 forward:

1. `internal/model/v41_config.go` — admit the loader identity (`"deepseek41"`) in
   `IsDeepSeekV41()`, keeping the safetensors `deepseek_v41` fail-closed gates intact.
2. `internal/model/config.go` — drop `"deepseek41"` from `usesMLAMoELayout()` so the
   non-MLA native branch is selected consistently (including `rope.go:45`'s partial-rotary
   denominator, which currently keys off the same predicate).
3. `internal/ggufload/deepseek41.go` — retarget the V4.1 per-layer suffix map from the
   glm-MLA names (`self_attn.q_a_proj/q_b_proj/kv_a_proj_with_mqa/o_proj_a/o_proj_b`,
   `mlp.shared_experts.*`, `attn.attn_sink`) to the native forward's names
   (`attn.wq_a/wq_b/wkv/wo_a/wo_b`, `attn.sink`, `ffn.shared_experts.w1/w3/w2`), matching
   `v41ForwardAdmitted`'s admitted shapes (`v41_forward.go:668-686`).
4. `internal/model/deepseek41_mla_route_test.go` — replace the false pin (it currently
   asserts `deepseek41.usesMLAMoELayout() == true`) with a pin that `deepseek41` routes to
   `forwardV41` and does **not** route to the GLM-DSA MLA forward.

## Gold-plating boundary

No roofline gate is lowered; no checkpoint is substituted; the `requireResidentShape`
fail-closed guard and the native v41 admission ladder stay green. Out of scope: any compute
HAL, kernel, or serving change; any change to the deepseek2 or GLM-5.2 MLA paths; the rope
convention itself (V4.1 applies the interleaved complex-pair rotation to the LAST
`rope_head_dim` of the head_dim row — `v41_forward.go:938` currently rotates the whole row
with `applyRopeRow`; that is a distinct numeric leaf and must be pinned separately against
the reference before this path can produce a token).

## Witness

`[SW-VERIFIED]` reference evidence (no hardware needed):

- Official reference, pinned revision, fetched 2026-09-17:
  `https://huggingface.co/deepseek-ai/DeepSeek-V4.1-Flash/raw/dba1be0a40aa45a94ad051997016db3960a90277/inference/model.py`
  (`class Attention`, `wkv = Linear(dim, head_dim)`, `kv_norm`, `apply_rotary_emb(kv[..., -rope_head_dim:])`).
- Official `config.json` at the same revision: `head_dim: 512`, `qk_rope_head_dim: 64`,
  `q_lora_rank: 1280`, `num_key_value_heads: 1`, no `kv_lora_rank`.
- Physical silicon `[HW-WITNESSED]` (build/warmup advance; NOT a token witness): the
  fak#13245 panic at `glm_dsa_session.go:443`, exclusive lease `lease-strix3-e40b7b7e74b6`,
  binary `b1261ee56aae2bd760ea09b5a80d6c61c7a1285831e0094f0b3c7a9b607f92b9`
  (`vcs.revision=9a4cfc973365`), log sha256
  `932626ca2e9570d9d88f9de4c134d31b6fa934b13cbc920c3544f59463b5ba4b`.

## Working spine

`Session.Step`/`Session.Prefill` dispatch on `IsDeepSeekV41()` -> `stepV41`/`prefillV41`
-> `forwardV41` -> `v41Layer` (`attn.wq_a/wq_b/wkv/wo_a/wo_b`, `attn.sink`, `mhc.*`,
`ffn.shared_experts.w1/w3/w2`). A `deepseek41` GGUF fails the predicate and falls to
`usesMLAMoELayout()` -> `tokenHiddenGLMDsa` -> `glmDsaAppendAttentionKV`
(`kv_a_proj_with_mqa`, `kv_b_proj`, `kv_a_layernorm`) -> the 576-wide request -> panic.

A loader identity and a model-side dispatch predicate must agree on which forward a
`deepseek41` artifact runs. Today they disagree in the worst way: the loader names the
family `deepseek41` and the MLA predicate claims it, while the native V4.1 predicate
(`IsDeepSeekV41`) does not recognize it. The correct through-line is to make the loader's
canonical naming and the native predicate agree, grounded in the official reference's
non-MLA `wkv` + `kv_norm` + partial in-place RoPE contract, so the exact artifact reaches
the forward that matches its tensors.

## Definition of done

Done condition: A `deepseek41` GGUF config routes to the native V4.1 forward and the
loader emits the native tensor names, so the exact vcruz Q2_K artifact no longer requests
a 576-wide `kv_a_proj_with_mqa` from a `[512,5120]` tensor.

- [ ] `[SW-VERIFIED]` `IsDeepSeekV41()` is true and `usesMLAMoELayout()` is false for
  `Config{ModelType: "deepseek41"}`, pinned by a regression test.
- [ ] `[SW-VERIFIED]` The suffix map emits the native names (`attn.wkv.weight`,
  `attn.wq_a.weight`, `attn.wq_b.weight`, `attn.wo_a.weight`, `attn.wo_b.weight`,
  `attn.sink`, `ffn.shared_experts.w1/w3/w2`), pinned by a loader test.
- [ ] `[SW-VERIFIED]` `go test ./internal/model/ ./internal/ggufload/ -count=1` GREEN,
  `go vet` exit 0, `gofmt -l` clean, 5-gate boundary PASS.
- [ ] `[HW-WITNESSED]` A physical strix3 serve advances past the `kv_a_proj_with_mqa`
  shape seam.

## Witnessed operating envelope

- host_available_gib: >= 54 GiB

## Likely files

- `internal/model/v41_config.go`
- `internal/model/config.go`
- `internal/ggufload/deepseek41.go`
- `internal/model/deepseek41_mla_route_test.go`
- `internal/model/kv.go` (dispatch call sites — likely no edit if (1)+(2) suffice)

## Acceptance gate

`go test ./internal/model/ -run 'V41|DeepSeek41|MLARoute' -count=1` and
`go test ./internal/ggufload/ -run DeepSeek41 -count=1` are GREEN; `go vet` exit 0;
`gofmt -l` clean; the 5-gate boundary PASS. Then the physical strix3 re-run.

## Closure binding

The resolving commit cites `anthony-chaudhary/fak#<this-issue>` and carries the
`(fak <leaf>)` trailer; the closing comment pastes the GREEN test output and the commit
sha in `anthony-chaudhary/fak`. No hardware witness is required to close the code leaf;
the physical re-run is tracked on the acceptance ledger and the physical-witness tracker.

## Parent context

Parent epic `anthony-chaudhary/fak#12640` (native DeepSeek V4.1 Flash execution and
physical qualification). Supersedes the framing in `anthony-chaudhary/fak#13245`
(keep #13245 as the physical re-run witness; this issue owns the code reconciliation).
Related: `anthony-chaudhary/fak#13244` (loader dialect hardening, distinct). The acceptance
ledger for this spine is `docs/goals/GOAL-deepseek-v41-halo-native.md` in the private
factory checkout that tracks this mission.

## Why this is next

Every loader-side blocker is closed and the forward is REACHED on silicon, but it is the
wrong forward: the mission has been debugging a GLM-MLA contract that V4.1 does not have.
Grounding the dispatch in the official reference converts an open-ended
projection-reconciliation chase into a bounded naming/dispatch leaf, and removes the
fixture-vs-artifact drift that produced both #13244 and #13245.

## Work estimate
Estimate: 3 points. Uncertainty: operator-supplied estimate; revise through the parent denominator when evidence changes.

## Overall completion contribution
Contribution: 3/24 points.

## Completion standard
production

## Target operating envelope
- host_available_gib: >= 50 GiB

**Comment by anthony-chaudhary** [comment]
Code leaf landed. `deepseek41` GGUF now routes to the native non-MLA V4.1 forward.

**Commit:** `5d931fa2f35c819af93c5041a710a218ffe8daba` on `origin/main` (public `anthony-chaudhary/fak`, pushed `d79bb8dbe..5d931fa2f`).

**What changed**
- `internal/model/v41_config.go` — `IsDeepSeekV41()` admits the loader identity `deepseek41`; `refuseDeepSeekV41Native()` admits it past every production load seam (`weights.go:374/458/477`, `safetensors.go`, AWQ/GPTQ/EXL2/quant loaders) while still refusing the safetensors published identity.
- `internal/model/config.go` — `usesMLAMoELayout()` drops `deepseek41`, so the non-MLA native branch is selected consistently.
- `internal/ggufload/deepseek41.go` — `deepseek41CanonicalSuffix` retargets the per-layer map to the native forward's admitted names (`attn.wq_a`/`wq_b`/`wkv`/`wo_a`/`wo_b`, `attn.sink`, `attn.kv_norm`, `indexer.*`, `attn.compressor.*`, `ffn.gate.*`, `ffn.shared_experts.w1/w3/w2`), matching `v41ForwardAdmitted` (`v41_forward.go:652-715`) and the reads at `:862-877`.
- `internal/model/deepseek41_mla_route_test.go`, `internal/ggufload/deepseek41_test.go` — regression pins.

**Definition-of-done evidence**
- `[SW-VERIFIED]` `IsDeepSeekV41()` true and `usesMLAMoELayout()` false for `Config{ModelType: "deepseek41"}` — pinned by `TestDeepSeek41IsDeepSeekV41` + `TestDeepSeek41DoesNotUseMLAMoELayout`.
- `[SW-VERIFIED]` Loader emits native names — pinned by `TestDeepSeek41GGUFV41RealArtifactSuffixMap`, `...FixtureResolvesEveryTensor`, `...FullNativeSuffixMap`.
- `[SW-VERIFIED]` Safetensors identity stays fail-closed — pinned by `TestDeepSeek41RefusalKeepsSafetensorsIdentityFailClosed` + `TestDeepSeek41GGUFIdentityAdmittedToNativeLoad`.
- `[SW-VERIFIED]` `go test ./internal/model/ ./internal/ggufload/ -count=1` GREEN; `go vet` exit 0; `gofmt -l` clean; 5-Gate boundary PASS (gates 0-5).
- Provenance: `Separation-Verdict: SEPARATED` (impl `router/mission-native-2026-09-18T034316Z`, test `worker/ses_f4d32f7fcffepi4Xka2FzqqBxb`).

**Witness output**
```
ok  github.com/anthony-chaudhary/fak/internal/model     6.329s
ok  github.com/anthony-chaudhary/fak/internal/ggufload  1.013s
```

**Retained invariants (not lowered)**
`requireResidentShape`, the native V4.1 admission ladder, and the safetensors `refuseDeepSeekV41Native` gate all stay fail-closed. No roofline gate lowered; the vcruz Q2_K artifact stays the pinned P1 artifact; the deepseek2/GLM MLA paths are untouched.

**Remaining frontier (not this leaf)**
The physical `[HW-WITNESSED]` re-run on strix3 — rebuild the on-device native `-tags vulkan` binary at `5d931fa2f` and re-serve the exact staged vcruz Q2_K artifact. The remaining expected seam is the rope convention: `v41_forward.go:938` rotates the whole head_dim row with `applyRopeRow`, while the reference applies the interleaved complex-pair rotation to only the LAST `qk_rope_head_dim` (64) of each head. That must be pinned separately against the official reference before this path can produce a token. Tracked on the acceptance ledger `docs/goals/GOAL-deepseek-v41-halo-native.md` and the physical-witness tracker `anthony-chaudhary/fak-private#1395`.

Parent epic: `anthony-chaudhary/fak#12640`. Supersedes the framing in `#13245` (which stays as the physical re-run witness).
